### PR TITLE
feat: Add fuzz testing for swap functionality, add rounding up (SC-482)

### PR DESCRIPTION
**Pull request by lucas-manuel**
Rounding up twice on sDai conversions because the rate conversion rounding down led to decreases in PSM value still.

### CODE REVIEW

File: test/unit/SwapExactIn.t.sol [hunk 1]
```diff
@@ -353,3 +353,118 @@ contract PSMSwapExactInSDaiAssetInTests is PSMSwapExactInSuccessTestsBase {
...
+            assertGe(psm.convertToAssetValue(psm.shares(lp0)), vars.lp0CachedValue);
+            assertGe(psm.convertToAssetValue(psm.shares(lp1)), vars.lp1CachedValue);
+            assertGe(psm.convertToAssetValue(psm.shares(lp2)), vars.lp2CachedValue);
+            assertGe(psm.totalAssets(),                        vars.psmCachedValue);
+
+            // Up to 2e12 rounding on each swap
+            assertApproxEqAbs(psm.convertToAssetValue(psm.shares(lp0)), vars.lp0CachedValue, 2e12);
+            assertApproxEqAbs(psm.convertToAssetValue(psm.shares(lp1)), vars.lp1CachedValue, 2e12);
+            assertApproxEqAbs(psm.convertToAssetValue(psm.shares(lp2)), vars.lp2CachedValue, 2e12);
+            assertApproxEqAbs(psm.totalAssets(),                   vars.psmCachedValue, 2e12);
```
Comment (hexonaut) [inline test/unit/SwapExactIn.t.sol:440]: Formatting.

File: test/unit/SwapExactIn.t.sol [hunk 2]
```diff
@@ -353,3 +353,118 @@ contract PSMSwapExactInSDaiAssetInTests is PSMSwapExactInSuccessTestsBase {
...
+        }
+
+        // Rounding is always in favour of the users
```
Comment (hexonaut) [inline test/unit/SwapExactIn.t.sol:443]: Rounding is always in favour of the LPs **

File: test/unit/SwapExactOut.t.sol [hunk 3]
```diff
@@ -344,7 +392,125 @@ contract PSMSwapExactOutSDaiAssetInTests is PSMSwapExactOutSuccessTestsBase {
...
+            vars.lp0CachedValue = psm.convertToAssetValue(psm.shares(lp0));
+            vars.lp1CachedValue = psm.convertToAssetValue(psm.shares(lp1));
+            vars.lp2CachedValue = psm.convertToAssetValue(psm.shares(lp2));
+            vars.psmCachedValue = psm.totalAssets();
+
+            assetIn.mint(swapper, amountIn);
+            assetIn.approve(address(psm), amountIn);
+            psm.swapExactOut(address(assetIn), address(assetOut), amountOut, amountIn, swapper, 0);
+
+            // Rounding is always in favour of the users
```
Comment (hexonaut) [inline test/unit/SwapExactOut.t.sol:476]: In favour of the LPs **

File: test/unit/SwapExactOut.t.sol [hunk 4]
```diff
@@ -344,7 +392,125 @@ contract PSMSwapExactOutSDaiAssetInTests is PSMSwapExactOutSuccessTestsBase {
...
+            assertGe(psm.totalAssets(),                        vars.psmCachedValue);
+
+            // Up to 2e12 rounding on each swap
+            assertApproxEqAbs(psm.convertToAssetValue(psm.shares(lp0)), vars.lp0CachedValue, 2e12);
+            assertApproxEqAbs(psm.convertToAssetValue(psm.shares(lp1)), vars.lp1CachedValue, 2e12);
+            assertApproxEqAbs(psm.convertToAssetValue(psm.shares(lp2)), vars.lp2CachedValue, 2e12);
+            assertApproxEqAbs(psm.totalAssets(),                        vars.psmCachedValue, 2e12);
+        }
+
+        // Rounding is always in favour of the users
```
Comment (hexonaut) [inline test/unit/SwapExactOut.t.sol:489]: LPs

File: test/unit/SwapExactIn.t.sol [hunk 5]
```diff
@@ -353,3 +353,118 @@ contract PSMSwapExactInSDaiAssetInTests is PSMSwapExactInSuccessTestsBase {
...
+            vars.lp0CachedValue = psm.convertToAssetValue(psm.shares(lp0));
+            vars.lp1CachedValue = psm.convertToAssetValue(psm.shares(lp1));
+            vars.lp2CachedValue = psm.convertToAssetValue(psm.shares(lp2));
+            vars.psmCachedValue = psm.totalAssets();
+
+            assetIn.mint(swapper, amountIn);
+            assetIn.approve(address(psm), amountIn);
+            psm.swapExactIn(address(assetIn), address(assetOut), amountIn, 0, swapper, 0);
+
+            // Rounding is always in favour of the users
```
Comment (hexonaut) [inline test/unit/SwapExactIn.t.sol:430]: LPs

**Review by hexonaut — CHANGES_REQUESTED** [review]
Looks good. Some minor changes.

File: test/unit/SwapExactIn.t.sol (hunk 5 above)
Comment (lucas-manuel) [reply test/unit/SwapExactIn.t.sol:430]: Updated

File: test/unit/SwapExactIn.t.sol (hunk 1 above)
Comment (lucas-manuel) [reply test/unit/SwapExactIn.t.sol:440]: Updated

File: test/unit/SwapExactIn.t.sol (hunk 2 above)
Comment (lucas-manuel) [reply test/unit/SwapExactIn.t.sol:443]: Updated

File: test/unit/SwapExactOut.t.sol (hunk 3 above)
Comment (lucas-manuel) [reply test/unit/SwapExactOut.t.sol:476]: Updated

File: test/unit/SwapExactOut.t.sol (hunk 4 above)
Comment (lucas-manuel) [reply test/unit/SwapExactOut.t.sol:489]: Updated

**Review by hexonaut — APPROVED** [review]

**Pull request merged**